apollo-router/tests/integration/subscriptions/mod.rs

```diff
@@ -175,60 +178,52 @@ pub async fn start_coprocessor_server() -> wiremock::MockServer {
 pub async fn verify_subscription_events(
     mut stream: impl futures::Stream<Item = Result<bytes::Bytes, reqwest::Error>> + Unpin,
     expected_events: Vec<serde_json::Value>,
+    include_heartbeats: bool,
 ) -> Result<Vec<serde_json::Value>, String> {
     use pretty_assertions::assert_eq;
 
     let mut subscription_events = Vec::new();
-
     // Set a longer timeout for receiving all events
     let timeout = tokio::time::timeout(tokio::time::Duration::from_secs(60), async {
+        let mut chunk_string = String::new();
         while let Some(chunk) = stream.next().await {
             let chunk = chunk.map_err(|e| format!("Stream error: {}", e))?;
-            let chunk_str = String::from_utf8_lossy(&chunk);
-
-            debug!("Received chunk: {}", chunk_str);
-
-            // Parse multipart chunks that contain GraphQL data
-            if chunk_str.contains("content-type: application/json") {
-                // Extract JSON from multipart response
-                if let Some(json_start) = chunk_str.find('{') {
-                    if let Some(json_end) = chunk_str.rfind('}') {
-                        let json_str = &chunk_str[json_start..=json_end];
-
-                        if let Ok(parsed) = serde_json::from_str::<serde_json::Value>(json_str) {
-                            // Store the raw parsed response without any transformation
-                            subscription_events.push(parsed.clone());
-                            info!(
-                                "Received subscription event {}: {}",
-                                subscription_events.len(),
-                                parsed
-                            );
-                        }
+            chunk_string = format!("{chunk_string}{}", String::from_utf8_lossy(&chunk));
+        }
+        let events = chunk_string
+            .split("\r\n--graphql\r\ncontent-type: application/json\r\n\r\n")
+            .filter_map(|s| {
+                let parsed = serde_json::from_str::<serde_json::Value>(
+                    s.trim_end_matches("\r\n--graphql--\r\n"), // If it's the last event
+                )
+                .ok()?;
+                if s.is_empty() {
+                    return None;
+                }
+                if parsed == serde_json::json!({}) {
+                    if include_heartbeats {
+                        Some(parsed)
+                    } else {
+                        None
                     }
+                } else {
+                    Some(parsed)
```

Comment:
to double check this is so we can use heartbeats in testing to verify subscriptions?

Reply:
Yes

Reply:
It's basically in the way we want to write tests if we want to include heartbeats or not